LoopKit/WeakObserverSet.swift

```diff
@@ -0,0 +1,147 @@
+            }
+        }
+    }
+
+    private let observers: Locked<[Identifier: Element]>
+
+    public init() {
+        observers = Locked([:])
+    }
+
+    /// Adds an observer and its calling queue
+    ///
+    /// - Parameters:
+    ///   - observer: The observer
+    ///   - queue: The queue to use when calling the observer
+    public func addObserver(_ observer: Observer, queue: DispatchQueue) {
+        insert(ObserverDispatchContainer(observer: observer, queue: queue))
+    }
+
+    /// Prunes any observer references that have been deallocated
+    /// - Returns: A reference to the instance for easy chaining
+    public func cleanupDeallocatedElements() -> Self {
+        observers.mutate { (storage) in
+            for (id, element) in storage where element.observer == nil {
+                storage.removeValue(forKey: id)
+            }
+        }
+        return self
+    }
+
+    /// Whether the observer is in the set
+    ///
+    /// - Parameter observer: The observer
+    /// - Returns: True if the observer is in the set
+    public func contains(_ observer: Observer) -> Bool {
+        let id = identifier(for: observer)
+        return observers.value[id] != nil
+    }
+
+    /// The total number of observers in the set
+    ///
+    /// Deallocated references are counted, so calling `cleanupDeallocatedElements` is advised to maintain accuracy of this value
+    public var count: Int {
+        return observers.value.count
+    }
+
+    /// Calls the given closure on each observer in the set, on the queue specified when the observer was added
+    ///
+    /// The order of calls is not defined
+    ///
+    /// - Parameter body: The closure to execute
+    public func forEach(_ body: @escaping (Observer) -> Void) {
+        // Hold the lock while we iterate, since each call is dispatched out
+        observers.mutate { (observers) in
+            observers.forEach { (pair) in
+                pair.value.call(body)
+            }
+        }
+    }
+
+    /// Removes the specified observer from the set
+    ///
+    /// - Parameter observer: The observer
+    public func removeObserver(_ observer: Observer) {
+        removeValue(forKey: identifier(for: observer))
+    }
+}
+
+extension WeakObserverSet {
+    private func identifier(for observer: Observer) -> ObjectIdentifier {
+        return ObjectIdentifier(observer as AnyObject)
+    }
+
+    private func identifier(for element: Element) -> ObjectIdentifier? {
+        guard let observer = element.observer else {
+            return nil
+        }
+        return identifier(for: observer)
+    }
+
+    @discardableResult
+    private func insert(_ newMember: Element) -> (inserted: Bool, memberAfterInsert: Element?) {
+        guard let id = identifier(for: newMember) else {
+            return (inserted: false, memberAfterInsert: nil)
+        }
+        var result: (inserted: Bool, memberAfterInsert: Element?)!
+        observers.mutate { (storage) in
```

Comment:
`mutate` returns the value returned by its closure, so the extra `result` variable can be dropped in favor of `return observers.mutate { ... }`

Reply:
`mutate()` returns the mutated value.

Reply:
Ah, oops. I'm thinking instead of a function with this type signature, which would make for a nice utility on Locked:
```swift
func withValue<Output>(_ transform: (T) -> Output) -> Output
```
which claims the lock, applies the transformation to the underlying value, and returns the output.